sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c

```diff
@@ -0,0 +1,304 @@
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
+    return rc;
+  }
+
+  mqtt_connect_options.username = (char*)az_span_ptr(username_span);
+  mqtt_connect_options.password = NULL; // Using X509 Client Certificate authentication.
+
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int subscribe()
+{
+  int rc;
+
+  char topic_filter[128];
+  az_span topic_filter_span = AZ_SPAN_LITERAL_FROM_BUFFER(topic_filter);
+  if ((rc = az_iot_provisioning_client_register_subscribe_topic_filter_get(
+           &provisioning_client, topic_filter_span, &topic_filter_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(topic_filter_span, '\0', &topic_filter_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_subscribe(mqtt_client, topic_filter, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int register_device()
+{
+  int rc;
+  MQTTClient_message pubmsg = MQTTClient_message_initializer;
+
+  char topic[128];
+  az_span topic_span = AZ_SPAN_LITERAL_FROM_BUFFER(topic);
+
+  if ((rc = az_iot_provisioning_client_register_publish_topic_get(
+           &provisioning_client, topic_span, &topic_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT PUB register topic, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(topic_span, '\0', &topic_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT PUB register topic, return code %d\n", rc);
+    return rc;
+  }
+
+  pubmsg.payload = NULL; // empty payload.
+  pubmsg.payloadlen = 0;
+  pubmsg.qos = 1;
+  pubmsg.retained = 0;
+
+  if ((rc = MQTTClient_publishMessage(mqtt_client, topic, &pubmsg, NULL)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish register request, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+int main()
+{
+  int rc;
+
+  if ((rc = read_configuration()) != AZ_OK)
+  {
+    printf("Failed to read configuration from environment variables, return code %d\n", rc);
+    return rc;
+  }
+
+  char client_id[128];
+  az_span client_id_span = AZ_SPAN_LITERAL_FROM_BUFFER(client_id);
+  if ((rc
+       = az_iot_provisioning_client_id_get(&provisioning_client, client_id_span, &client_id_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(client_id_span, '\0', &client_id_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
```

Comment:
not exactly the error here

Reply:
Same as above - I won't change before we decide on the public API.